File: backend/src/main/java/com/workflow/engine/form/FormDataService.java
package com.workflow.engine.form;

import com.workflow.engine.form.entity.FormData;
import com.workflow.engine.form.entity.FormDefinition;
import com.workflow.engine.form.repository.FormDataRepository;
import com.workflow.engine.form.repository.FormDefinitionRepository;
import com.workflow.engine.tenant.TenantProvider;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * 表单实例数据服务。
 * 管理流程实例关联的表单数据的保存和查询。
 *
 * 版本快照：保存时记录当前表单定义的版本号（form_version），
 * 保证旧数据与旧 schema 对应。
 */
@Service
public class FormDataService {

    private final FormDataRepository formDataRepository;
    private final FormDefinitionRepository formDefRepository;
    private final TenantProvider tenantProvider;

    public FormDataService(FormDataRepository formDataRepository,
                           FormDefinitionRepository formDefRepository,
                           TenantProvider tenantProvider) {
        this.formDataRepository = formDataRepository;
        this.formDefRepository = formDefRepository;
        this.tenantProvider = tenantProvider;
    }

    /**
     * 保存或更新当前表单数据（非快照，用于节点间传递）。
     * 同一 processInstanceId + formDefId 只保留一条当前数据。
     *
     * @param formDefId          表单定义 ID
     * @param processInstanceId  流程实例 ID
     * @param taskId             任务 ID（可选）
     * @param dataJson           表单数据 JSON
     * @return 创建或更新的表单数据记录
     */
    @Transactional
    public FormData save(String formDefId, String processInstanceId, String taskId, String dataJson) {
        String tenantId = tenantProvider.getTenantId();

        // 获取表单定义的当前版本作为快照
        FormDefinition formDef = formDefRepository.findByIdAndTenantId(formDefId, tenantId)
                .orElseThrow(() -> new RuntimeException("Form definition not found: " + formDefId));

        // 查找是否已有当前数据，有则更新，无则创建
        Optional<FormData> existing = formDataRepository
                .findByTenantIdAndProcessInstanceIdAndFormDefIdAndIsSnapshot(
                        tenantId, processInstanceId, formDefId, false);

        FormData formData;
        if (existing.isPresent()) {
            formData = existing.get();
            formData.setDataJson(dataJson);
            formData.setTaskId(taskId);
            formData.setFormVersion(formDef.getVersion());
        } else {
            formData = new FormData();
            formData.setId(UUID.randomUUID().toString().replace("-", ""));
            formData.setTenantId(tenantId);
            formData.setFormDefId(formDefId);
            formData.setFormVersion(formDef.getVersion());
            formData.setProcessInstanceId(processInstanceId);
            formData.setTaskId(taskId);
            formData.setDataJson(dataJson);
            formData.setIsSnapshot(false);
        }

        return formDataRepository.save(formData);
    }

    /**
     * 保存任务审批时的表单快照（每次创建新记录，不可变）。
     *
     * @param formDefId          表单定义 ID
     * @param processInstanceId  流程实例 ID
     * @param taskId             任务 ID
     * @param dataJson           表单数据 JSON
     * @return 创建的快照记录
     */
    @Transactional
    public FormData saveSnapshot(String formDefId, String processInstanceId, String taskId, String dataJson) {
        String tenantId = tenantProvider.getTenantId();

        FormDefinition formDef = formDefRepository.findByIdAndTenantId(formDefId, tenantId)
                .orElseThrow(() -> new RuntimeException("Form definition not found: " + formDefId));

        FormData snapshot = new FormData();
        snapshot.setId(UUID.randomUUID().toString().replace("-", ""));
        snapshot.setTenantId(tenantId);
        snapshot.setFormDefId(formDefId);
        snapshot.setFormVersion(formDef.getVersion());
        snapshot.setProcessInstanceId(processInstanceId);
        snapshot.setTaskId(taskId);
        snapshot.setDataJson(dataJson);
        snapshot.setIsSnapshot(true);

        return formDataRepository.save(snapshot);
    }

    /**
     * 获取单条表单数据。
     */
    public FormData getById(String id) {
        String tenantId = tenantProvider.getTenantId();
        return formDataRepository.findByIdAndTenantId(id, tenantId)
                .orElseThrow(() -> new RuntimeException("Form data not found: " + id));
    }

    /**
     * 按流程实例和表单定义查询当前表单数据（非快照）。
     */
    public Optional<FormData> findByProcessInstance(String processInstanceId, String formDefId) {
        String tenantId = tenantProvider.getTenantId();
        return formDataRepository
                .findByTenantIdAndProcessInstanceIdAndFormDefIdAndIsSnapshot(
                        tenantId, processInstanceId, formDefId, false);
    }

    /**
     * 按 taskId 查询审批快照。
     */
    public Optional<FormData> findByTaskId(String taskId) {
        String tenantId = tenantProvider.getTenantId();
        return formDataRepository
                .findByTenantIdAndTaskIdAndIsSnapshotOrderByCreatedAtDesc(tenantId, taskId, true)
                .stream()
                .findFirst();
    }

    /**
     * 按流程实例查询所有表单数据（含快照）。
     */
    public List<FormData> findByProcessInstance(String processInstanceId) {
        String tenantId = tenantProvider.getTenantId();
        return formDataRepository
                .findByTenantIdAndProcessInstanceId(tenantId, processInstanceId);
    }

    /**
     * 按流程实例查询所有审批快照（按时间倒序）。
     */
    public List<FormData> findSnapshotsByProcessInstance(String processInstanceId) {
        String tenantId = tenantProvider.getTenantId();
        return formDataRepository
                .findByTenantIdAndProcessInstanceIdAndIsSnapshotOrderByCreatedAtDesc(
                        tenantId, processInstanceId, true);
    }

    /**
     * 更新当前表单数据（非快照）。
     */
    @Transactional
    public FormData update(String id, String dataJson) {
        String tenantId = tenantProvider.getTenantId();
        FormData formData = formDataRepository.findByIdAndTenantId(id, tenantId)
                .orElseThrow(() -> new RuntimeException("Form data not found: " + id));

        formData.setDataJson(dataJson);

        return formDataRepository.save(formData);
    }

    /**
     * 保存发起页草稿（processInstanceId 为 null 的表单数据）。
     * 同一 formDefId 只保留一条草稿，再次保存为更新。
     *
     * @param formDefId 表单定义 ID
     * @param dataJson  表单数据 JSON
     * @return 保存的草稿记录
     */
    @Transactional
    public FormData saveDraft(String formDefId, String dataJson) {
        String tenantId = tenantProvider.getTenantId();

        FormDefinition formDef = formDefRepository.findByIdAndTenantId(formDefId, tenantId)
                .orElseThrow(() -> new RuntimeException("Form definition not found: " + formDefId));

        Optional<FormData> existing = formDataRepository
                .findByTenantIdAndFormDefIdAndProcessInstanceIdIsNullAndIsSnapshot(tenantId, formDefId, false);

        FormData formData;
        if (existing.isPresent()) {
            formData = existing.get();
            formData.setDataJson(dataJson);
            formData.setFormVersion(formDef.getVersion());
        } else {
            formData = new FormData();
            formData.setId(UUID.randomUUID().toString().replace("-", ""));
            formData.setTenantId(tenantId);
            formData.setFormDefId(formDefId);
            formData.setFormVersion(formDef.getVersion());
            formData.setProcessInstanceId(null);
            formData.setTaskId(null);
            formData.setDataJson(dataJson);
            formData.setIsSnapshot(false);
        }

        return formDataRepository.save(formData);
    }

    /**
     * 查询发起页草稿。
     *
     * @param formDefId 表单定义 ID
     * @return 草稿记录，无草稿返回 empty
     */
    public Optional<FormData> findDraft(String formDefId) {
        String tenantId = tenantProvider.getTenantId();
        return formDataRepository
                .findByTenantIdAndFormDefIdAndProcessInstanceIdIsNullAndIsSnapshot(tenantId, formDefId, false);
    }

    /**
     * 清除发起页草稿（发起成功后调用）。
     *
     * @param formDefId 表单定义 ID
     */
    @Transactional
    public void clearDraft(String formDefId) {
        String tenantId = tenantProvider.getTenantId();
        formDataRepository
                .findByTenantIdAndFormDefIdAndProcessInstanceIdIsNullAndIsSnapshot(tenantId, formDefId, false)
                .ifPresent(formDataRepository::delete);
    }
}

File: backend-node/src/engine/form/form-data.service.ts
import { Injectable } from '@nestjs/common'
import { randomBytes } from 'node:crypto'
import { bitToBool } from '../../framework/database/types'
import { getTenantId } from '../../framework/tenant/tenant-context'
import { FormDefinitionRepository } from './repository/form-definition.repository'
import {
  FormDataRepository,
  type FormDataRow,
} from './repository/form-data.repository'

/**
 * 表单实例数据服务，对齐 Java `com.workflow.engine.form.FormDataService`。
 *
 * 三个「同一张表、三种语义」的概念，别混：
 *   - **当前数据**：同一 `进程实例 + 表单定义` 只有一条，保存时 upsert
 *   - **审批快照**：每次保存都新建，不可变
 *   - **发起页草稿**：`process_instance_id IS NULL` 的那条，保存时 upsert
 *
 * ⚠️ 错误类型：Java 这里一律用 `RuntimeException`（表单定义/数据不存在），
 *    兜底异常处理器映射成 **HTTP 500**，不是 200 + body 内 code。
 *    所以下面用普通 `Error` 而不是 `BusinessException` —— 后者会改变 HTTP 状态码。
 */
@Injectable()
export class FormDataService {
  constructor(
    private readonly repository: FormDataRepository,
    private readonly formDefRepository: FormDefinitionRepository,
  ) {}

  /** 保存或更新当前数据（upsert）。 */
  async save(
    formDefId: string | null,
    processInstanceId: string | null,
    taskId: string | null,
    dataJson: string | null,
  ): Promise<FormDataVO> {
    const tenantId = getTenantId()
    const formVersion = await this.requireFormVersion(formDefId, tenantId)
    const existing = await this.repository.findCurrent(
      tenantId,
      String(processInstanceId),
      String(formDefId),
    )

    if (existing !== null) {
      await this.repository.updateData(existing.id, String(dataJson ?? ''))
      const updated = await this.repository.findByIdAndTenantId(existing.id, tenantId)
      return toVO(updated ?? existing)
    }

    const row: FormDataRow = {
      id: newId(),
      tenant_id: tenantId,
      form_def_id: String(formDefId),
      form_version: formVersion,
      process_instance_id: processInstanceId,
      task_id: taskId,
      data_json: dataJson,
      created_by: null,
      created_at: new Date(),
      updated_at: new Date(),
      is_snapshot: 0,
    }
    await this.repository.insert(row)
    return toVO(row)
  }

  /** 保存审批快照（每次新建，不可变）。 */
  async saveSnapshot(
    formDefId: string | null,
    processInstanceId: string | null,
    taskId: string | null,
    dataJson: string | null,
  ): Promise<FormDataVO> {
    const tenantId = getTenantId()
    const formVersion = await this.requireFormVersion(formDefId, tenantId)
    const row: FormDataRow = {
      id: newId(),
      tenant_id: tenantId,
      form_def_id: String(formDefId),
      form_version: formVersion,
      process_instance_id: processInstanceId,
      task_id: taskId,
      data_json: dataJson,
      created_by: null,
      created_at: new Date(),
      updated_at: new Date(),
      // BIT(1)：显式写 0/1，别依赖驱动把 boolean 转成 bit
      is_snapshot: 1,
    }
    await this.repository.insert(row)
    return toVO(row)
  }

  /** 保存发起页草稿（upsert，实例为 NULL）。 */
  async saveDraft(formDefId: string | null, dataJson: string | null): Promise<FormDataVO> {
    const tenantId = getTenantId()
    const formVersion = await this.requireFormVersion(formDefId, tenantId)
    const existing = await this.repository.findDraft(tenantId, String(formDefId))

    if (existing !== null) {
      await this.repository.updateData(existing.id, String(dataJson ?? ''))
      const updated = await this.repository.findByIdAndTenantId(existing.id, tenantId)
      return toVO(updated ?? existing)
    }

    const row: FormDataRow = {
      id: newId(),
      tenant_id: tenantId,
      form_def_id: String(formDefId),
      form_version: formVersion,
      process_instance_id: null,
      task_id: null,
      data_json: dataJson,
      created_by: null,
      created_at: new Date(),
      updated_at: new Date(),
      is_snapshot: 0,
    }
    await this.repository.insert(row)
    return toVO(row)
  }

  /** 按 id 取单条；不存在抛 500（对齐 Java 的 RuntimeException）。 */
  async getById(id: string): Promise<FormDataDTOVO> {
    const row = await this.repository.findByIdAndTenantId(id, getTenantId())
    if (row === null) throw new Error(`Form data not found: ${id}`)
    return toDTO(row)
  }

  /** 按实例 + 表单定义取当前数据（无则 null）。 */
  async findByProcessInstance(
    processInstanceId: string,
    formDefId: string,
  ): Promise<FormDataDTOVO | null> {
    const row = await this.repository.findCurrent(
      getTenantId(),
      processInstanceId,
      formDefId,
    )
    return row === null ? null : toDTO(row)
  }

  /** 按 taskId 取最新审批快照（无则 null）。 */
  async findByTaskId(taskId: string): Promise<FormDataDTOVO | null> {
    const row = await this.repository.findSnapshotByTaskId(getTenantId(), taskId)
    return row === null ? null : toDTO(row)
  }

  /** 某实例的全部表单数据（含快照）。 */
  async findByProcessInstanceAll(processInstanceId: string): Promise<FormDataDTOVO[]> {
    const rows = await this.repository.findByProcessInstance(getTenantId(), processInstanceId)
    return rows.map(toDTO)
  }

  /** 某实例的全部审批快照（创建时间倒序）。 */
  async findSnapshots(processInstanceId: string): Promise<FormDataDTOVO[]> {
    const rows = await this.repository.findSnapshots(getTenantId(), processInstanceId)
    return rows.map(toDTO)
  }

  /** 查询发起页草稿（无则 null）。 */
  async findDraft(formDefId: string): Promise<FormDataDTOVO | null> {
    const row = await this.repository.findDraft(getTenantId(), formDefId)
    return row === null ? null : toDTO(row)
  }

  /** 清除发起页草稿（不存在时静默成功）。 */
  async clearDraft(formDefId: string): Promise<void> {
    const row = await this.repository.findDraft(getTenantId(), formDefId)
    if (row !== null) await this.repository.deleteById(row.id)
  }

  /** 更新当前数据（不动 form_version / task_id，对齐 Java）。 */
  async update(id: string, dataJson: string | null): Promise<FormDataVO> {
    const tenantId = getTenantId()
    const row = await this.repository.findByIdAndTenantId(id, tenantId)
    if (row === null) throw new Error(`Form data not found: ${id}`)
    await this.repository.updateData(id, String(dataJson ?? ''))
    const updated = await this.repository.findByIdAndTenantId(id, tenantId)
    return toVO(updated ?? row)
  }

  /** 取表单定义的当前版本（用于写快照版本号）；不存在抛 500。 */
  private async requireFormVersion(
    formDefId: string | null,
    tenantId: string,
  ): Promise<number> {
    const row = await this.formDefRepository.findByIdAndTenantId(String(formDefId), tenantId)
    if (row === null) throw new Error(`Form definition not found: ${String(formDefId)}`)
    return row.version
  }
}

/** 对齐 Java 的 ID 生成：UUID 去掉横线 → 32 位十六进制。 */
function newId(): string {
  return randomBytes(16).toString('hex')
}

/**
 * 写端点返回的形状 —— Java 直接返回 **`FormData` 实体**（含 `tenantId`）。
 */
export interface FormDataVO {
  id: string
  tenantId: string
  formDefId: string
  formVersion: number
  processInstanceId: string | null
  taskId: string | null
  dataJson: string | null
  createdBy: string | null
  createdAt: Date | null
  updatedAt: Date | null
  isSnapshot: boolean
}

/** 读端点返回的形状 —— Java 返回 **`FormDataDTO`**（**没有** `tenantId`）。 */
export interface FormDataDTOVO {
  id: string
  formDefId: string
  formVersion: number
  processInstanceId: string | null
  taskId: string | null
  dataJson: string | null
  createdBy: string | null
  createdAt: Date | null
  updatedAt: Date | null
  isSnapshot: boolean
}

function toVO(row: FormDataRow): FormDataVO {
  return {
    id: row.id,
    tenantId: row.tenant_id,
    formDefId: row.form_def_id,
    formVersion: row.form_version,
    processInstanceId: row.process_instance_id,
    taskId: row.task_id,
    dataJson: row.data_json,
    createdBy: row.created_by,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
    isSnapshot: bitToBool(row.is_snapshot),
  }
}

function toDTO(row: FormDataRow): FormDataDTOVO {
  return {
    id: row.id,
    formDefId: row.form_def_id,
    formVersion: row.form_version,
    processInstanceId: row.process_instance_id,
    taskId: row.task_id,
    dataJson: row.data_json,
    createdBy: row.created_by,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
    isSnapshot: bitToBool(row.is_snapshot),
  }
}
